compiler/expressions.go

```diff
@@ -969,7 +981,12 @@ func (fc *funcContext) translateBuiltin(name string, sig *types.Signature, args
 	case "delete":
 		args = fc.expandTupleArgs(args)
 		keyType := fc.pkgCtx.TypeOf(args[0]).Underlying().(*types.Map).Key()
-		return fc.formatExpr(`delete %e[%s.keyFor(%s)]`, args[0], fc.typeName(keyType), fc.translateImplicitConversion(args[1], keyType))
+		return fc.formatExpr(
+			`(typeof %1e.delete === "function" && %1e.delete(%2s.keyFor(%3s)))`,
```

Comment:
When would delete not be a function? And what should we do in such a case?

Reply:
I don't know of a more graceful way of handling this, but it is possible to call `delete(nil)` in go and the javascript would wind up calling delete() on undefined.  I'm hand waving a bit based on what I saw.  I could make an explicit test case, if that would help?

Reply:
Ah. If that's the only edge case, then this is probably correct. Although perhaps checking for null/js.Undefined would be more explicit, if that's not too cumbersome.

Reply:
I was worried that there might be some situations where a JS object had been set, but it wouldn't acutally be a Map type, and an undefined check would still cause an error.  Perhaps that is an unfounded worry.

Reply:
That's the kind of corner case I was wondering about. If that were to happen, what would it actually mean?

Reply:
As long as the places that gopherjs compiler makes Maps are all using `new Map()`, I would expect things would be generally good.  AFAIK, I got all those.  The edge cases would operate outside of that object flow.  I'm not sure what sort of ways that could happen.  

Passing in a wrong type is a RTE.  Maybe there is a way to coerce a type so that delete() is satisfied, but the actual object is incorrect?  Maybe using `js.Global` to make an object and coerce that into a map to pass to delete?

Reply:
For the test I just committed, without the typeof protection, delete called on a nil map would result in this:
```
--- FAIL: Test_MapDelete (0.00s)
TypeError: nilMap.delete is not a function
    at Test_MapDelete (/github.com/gopherjs/gopherjs/tests/map_test.go:140:3)
    at tRunner (/testing/testing.go:1439:3)
    at $goroutine (/Users/tomconnell/go/src/github.com/gopherjs/gopherjs/tests/tests_test.3034384146.js:1632:19)
    at $runScheduled (/Users/tomconnell/go/src/github.com/gopherjs/gopherjs/tests/tests_test.3034384146.js:1678:7)
    at $schedule (/Users/tomconnell/go/src/github.com/gopherjs/gopherjs/tests/tests_test.3034384146.js:1702:5)
    at queueEntry (/Users/tomconnell/go/src/github.com/gopherjs/gopherjs/tests/tests_test.3034384146.js:1775:5)
    at $send (/Users/tomconnell/go/src/github.com/gopherjs/gopherjs/tests/tests_test.3034384146.js:1734:5)
    at $b (/Users/tomconnell/go/src/github.com/gopherjs/gopherjs/tests/tests_test.3034384146.js:7687:9)
    at $b (/syscall/js/gopherjs__js.go:83:5)
    at /Users/tomconnell/go/src/github.com/gopherjs/gopherjs/tests/tests_test.3034384146.js:68:72
    at FSReqCallback.wrapper [as oncomplete] (node:fs:798:5)
FAIL	github.com/gopherjs/gopherjs/tests	0.719s
```

In other words, I believe the code is correct.